rfcs/20200612-stream-executor-c-api.md

```diff
@@ -0,0 +1,597 @@
+  // Must be filled by the plugin.
+  SP_Platform platform;  // out
+} SE_PlatformRegistrationParams;
+
+#define SE_PLATFORM_REGISTRATION_PARAMS_SIZE TF_OFFSET_OF_END(SE_PlatformRegistrationParams, platform)
+
+void SE_InitializePlugin(SE_PlatformRegistrationParams* params, TF_Status* status);
+
+#ifdef __cplusplus
+} // extern "C"
+#endif
+```
+
+## Registration implementation
+
+Registration will be implemented by registering a new StreamExecutor platform as well as a new TensorFlow device with [DeviceFactory](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/common_runtime/device_factory.h;l=30?q=DeviceFactory).
+
+```cpp
+typedef (*SEPluginInitFn)(SE_PlatformRegistrationParams*, TF_Status*);
+...
+
+void* plugin = dlopen("myplugin.so", ...);
+if (!plugin) {
+  ... output error and skip this plugin ...
+}
+void* initialize_sym = dlsym(plugin, "SE_InitializePlugin");
+if (!initialize_sym) {
+  ... output error and skip this plugin ...
+}
+SEPluginInitFn initialize_fn = reinterpret_cast<SEPluginInitFn>(initialize_sym);
+
+SE_PlatformRegistrationParams params;
+TF_Status* status = TF_NewStatus();
+
+initialize_fn(&params, status);
+   
+// Register new platform
+std::unique_ptr<stream_executor::internal::CPlatform> platform(
+    new stream_executor::internal::CPlatform(params));
+SE_CHECK_OK(
+   stream_executor::MultiPlatformManager::RegisterPlatform(
+    std::move(platform)));
+   
+// Register PluggableDevice
+std::string platform_name_str(params.params.name, params.params.name_len);
+std::string type_str(params.params.type, params.params.type_len);
+DeviceFactory::Register(type_str, new PluggableDeviceFactory(platform_name_str), priority);
+```
+
+`PluggableDevice` is covered in a separate RFC: [RFC: Adding Pluggable Device For TensorFlow](https://github.com/tensorflow/community/pull/262).
+
+## PlatformId
+
+StreamExecutor [Platform](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/platform.h;l=114) has an id parameter. This parameter will be hidden from the C API and set internally by TensorFlow instead.
+
+## Usage example
+
+Define functions that create and destroy `SE_Device` and `SE_StreamExecutor`:
+
+```cpp
+void create_device(SP_Device* device, SE_Options* options, TF_Status* status) {
+  device->device_handle = get_my_device_handle();
+  ...
+}
+void create_stream_executor(SP_StreamExecutor* se, TF_Status* status) {
+  se->memcpy_from_host = my_device_memcpy_from_host_function;
+  ...
+}
+void destroy_device(SP_Device* device) {
+  -- destroy device handle here --
+  ...
+}
+void destroy_stream_executor(SP_StreamExecutor* stream_executor) {
+  -- perform any clean up needed for stream executor --
+}
+```
+
+Define `SE_InitializePlugin` that TensorFlow will call when registering the device plugin:
+
+```cpp
+void SE_InitializePlugin(SE_PlatformRegistrationParams* params, TF_Status* status) {
+  int32_t visible_device_count = 2;
+  
+  std::string name = "MyDevice";
```

Comment:
@annarev  do you prefer to add a new parameter  for sub_device_type or to share the the platform name with the sub device type. Seems both sub device type and platform name needs to be different between plugins(one maps to kernel and one maps to streamexecutor platform), so I am wondering whether they can be the same thing. Thanks

Reply:
Hm I am thinking we can use the same name. Can you think of a case when we need different platform vs sub-device type?

(sorry for delay in reply)

Reply:
I also think they can share the same string since one plugin should have a unique platform name and a unique sub-device type. I didn't think out a case that we need a platform name with multiple sub-device types.:)